src/Microsoft.ML.Tokenizers/Model/SentencePieceBpe.cs

```diff
@@ -0,0 +1,952 @@
+                {
+                    continue;
+                }
+
+                // Replaces symbols with `top` rule.
+                symbols[top.Left].pieceSpan = (symbols[top.Left].pieceSpan.Index, symbols[top.Left].pieceSpan.Length + symbols[top.Right].pieceSpan.Length);
+                symbols[top.Left].id = top.Id;
+
+                // Updates prev/next pointers.
+                symbols[top.Left].next = symbols[top.Right].next;
+
+                if (symbols[top.Right].next >= 0)
+                {
+                    symbols[symbols[top.Right].next].prev = top.Left;
+                }
+                symbols[top.Right].pieceSpan = (0, 0);
+
+                // Adds new symbol pairs which are newly added after symbol replacement.
+                TryMerge(symbols[top.Left].prev, top.Left, text);
+                TryMerge(top.Left, symbols[top.Left].next, text);
+            }
+
+            return revMerge;
+
+            void TryMerge(int left, int right, ReadOnlySpan<char> textSpan)
+            {
+                if (left == -1 || right == -1)
+                {
+                    return;
+                }
+
+                int pieceLength = symbols[left].pieceSpan.Length + symbols[right].pieceSpan.Length;
+                if (!_vocab.TryGetValue(textSpan.Slice(symbols[left].pieceSpan.Index, pieceLength), out (int Id, float Score, byte Type) leftId))
+                {
+                    return;
+                }
+
+                symbols[left].type = leftId.Type;
+
+                SymbolPair pair = new(left, right, leftId.Score, pieceLength, leftId.Id);
+                agenda.Enqueue(pair);
+
+                if (leftId.Type == (byte)ModelProto.Types.SentencePiece.Types.Type.Unused)
+                {
+                    revMerge ??= new();
+                    revMerge.Add((symbols[left].pieceSpan.Index, pieceLength), (symbols[left].pieceSpan.Index, symbols[left].pieceSpan.Length, symbols[right].pieceSpan.Index, symbols[right].pieceSpan.Length));
+                }
+            }
+        }
+
+        private struct SymbolPair : IEquatable<SymbolPair>, IComparable<SymbolPair>
+        {
+            public int Left { get; set; }
+            public int Right { get; set; }
+            public int Length { get; set; }
+            public float Score { get; set; }
+            public int Id { get; set; }
+
+            public SymbolPair(int left, int right, float score, int length, int id)
+            {
+                Left = left;
+                Right = right;
+                Score = score;
+                Length = length;
+                Id = id;
+            }
+
+            public int CompareTo(SymbolPair other)
+            {
+                if (Score != other.Score)
+                {
+                    return other.Score.CompareTo(Score);
+                }
+
+                return other.Left.CompareTo(Left);
```

Comment:
Right, Length, and Id don't need to factor into comparisons?

Reply:
No the score and Left is enough. basically the items stored in the queue are part of a linked list. if the Left is the same, this means it is same item inside the linked list.